### ISSUE TITLE
Support saving and loading event snapshots

### ISSUE THREAD

**Issue by ojura** [mozilla/rr#1989]
This feature request was inspired by a scenario in which I had to debug an application where the bug occurs after a long time into the execution (in my example 20 minutes). I figured out the exact event number (around 600 000). Each time when I want to start debugging, I start rr with `-g 600000` and go grab a cup of coffee until the application execution reaches the interesting event. Once rr reaches this event, I can normally use forward and backward execution to do debugging. The kicker is, I can always restart to the interesting event with `run 600000`, and I won't have to wait for long because rr can intelligently rewind to this event.

The feature request is the following: could I save the state of the application at the 600 000th event to a snapshot, and tell rr to load that snapshot on startup instead of waiting for rr to run the application from beginning?

Better yet, rr could maintain a list of saved snapshots in a trace, and automatically choose the closest one when calling `run` in order to minimize the time that the user will have to wait for the execution to reach that event. The workflow would look something like this:

```
# this takes 20 minutes
$ rr record app

# this also takes 20 minutes
$ rr replay -g 600000

(rr) save snapshot
Created snapshot at event 600000.
(rr) quit

$ rr replay -g 600000 
# rr figures out that it has a snapshot at this time,
# so this is instantaneous

# ...do debugging...
(rr) quit

# The next time we start a debugging session, we decide 
# to go a little further back in time:
$ rr replay -g 550000 
# this is also significantly faster, because rr chooses 
# the nearest snapshot (in this case 600000), and replays
# backwards from there
```

**Comment by rocallahan** [comment]
rr replay already knows how to create in-memory checkpoints and resume from them. So this task can be split into a few steps:
* Support saving a replay checkpoint to stable storage
* Support loading a replay checkpoint from stable storage
* Support saving a replay checkpoint during recording

Currently replay checkpoints are stored in a `CloneCompletion` in `Session::copy_state_to`. It's a `vector` of `TaskGroup`s. For each `TaskGroup` we select one task, the `clone_leader` which is `fork`ed from the original task. This duplicates the address space and all the private mappings "for free". Other tasks in the `TaskGroup` are created lazily. All the information needed to reconstitute each task is stored in a `CapturedState` per `Task`.

(That setup is slightly wrong. A `TaskGroup` represents a Linux thread group (should be renamed) but different thread groups can share an address space (e.g. during vfork), and really there should be one `clone_leader` per `AddressSpace`. That should probably be fixed first.)

Then to save a replay checkpoint, one would have to write out the `CapturedState`s (easy) and also iterate through the maps in the `clone_leader`'s `AddressSpace` and write out their contents if necessary. Determining exactly when that's necessary is not easy :-(. For example, even pages of private read-only mappings of non-writeable files may have been modified via /proc/<pid>/mem or `mprotect` and need to be saved somewhere. `/proc/.../mapfiles` may help. Just writing out all pages (and compressing) would be easy but produce large dumps. Maybe you should compare the contents of pages with the underlying mapped files to see what's changed.

Saving a checkpoint during recording might just be able to use the same or very similar code. The key thing is that we only allow replay from a checkpoint, not arbitrary execution, so we don't have to do everything [http://criu.org](CRIU).

**Comment by rocallahan** [comment]
It would probably make sense to store checkpoint data in the trace directory in subdirectories, e.g. `checkpoint-0`, `checkpoint-1` etc. With some metadata to indicate the current event number, tick count, register values, and tid.

**Comment by ojura** [comment]
I would like to add that the description of this is completely wrong for the second part, where we replay to event 550000; we can not quickly rewind there from the event 600000 if we don't have any checkpoints stored before the event 550000. At the time of proposing, I did not fully understand how rr actually would rewind in this case (and that is by finding an earlier checkpoint and running forward from there).

**Comment by alexeypa** [comment]
Another use case for checkpoints is storing the trace in a circular buffer. I.e. rr could create a checkpoint every now and then and discard the trace data accumulated before the checkpoint. This will allow replay from the last checkpoint saving 20 minutes spent on `rr replay -g 600000` in the scenario above. This is useful if we want to run the trace for hours until certain interesting event occurs.

**Comment by ojura** [comment]
@alexeypa, doesn't rr already support this? It already creates internal checkpoints which are used for backwards execution. This issue is more related to saving these checkpoints to a file, so you don't have to wait until you reach them for the first time.

**Comment by alexeypa** [comment]
> This issue is more related to saving these checkpoints to a file, so you don't have to wait until you reach them for the first time.

No, I understand that. What I meant is a step forward from that. Not only checkouts are stored in a file, but the trace data preceding the last checkpoint is purged (when rr configured this way) from the trace. This still allows to replay the most interesting part of the trace leading to the crash but keeps the trace small to allow recording to go for a long time.

This should probably be filed as a separate issue.

**Comment by ojura** [comment]
Ah, I understand - storing the trace in a circular buffer to limit its size. And this depends on implementing the functionality described here, since we have to start the replay from the first checkpoint in the circular buffer (instead of the beginning of the trace).

**Comment by alexeypa** [comment]
Yes,  exactly.

**Comment by albertvucinovic** [comment]
Is there some progress on the issue?
Also, a solution could be to use criu for checkpointing.
I have tried that, but it seams that criu can not checkpoint rr.
It seams to me that is because of some bug which is related to using ptrace and SIGCHLD in a composed fashion?
Would that be a rr bug or criu bug, or is combining criu with rr principally impossible?

**Comment by rocallahan** [comment]
No-one has worked on this AFAIK.

criu isn't going to be helpful here because we need rr to manage the checkpoints.

**Comment by albertvucinovic** [comment]
Thank you very much for your answer, I did have a misunderstanding of rr's checkpoints and what you need them for!
What would be the easiest way to spawn a new process which would continue from a checkpoint created by the rr gdb backend (or simple from a specific point in the trace), but not in the context of the recorded trace, but as a live process?

**Comment by rocallahan** [comment]
That is not possible in general because the kernel state (which files are open etc) is not present in the rr replay.
